### PR TITLE
[RDY] Allow diagonal movement via keybinding modifiers in SDL builds

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
First approach for #20447.

 # What was changed:

* allows diagonal movement with cursor keys using CTRL and SHIFT modifiers via `DIAG_MOVE_WITH_MODIFIERS` interface option (_currently enabled only in SDL builds and disabled in CURSES build_);
* diagonal movement action keys are taken from keybindings, so you need these to be configured for diagonal movements to work.

# Default key bindings for diagonal movement:

- `1` = `Move Southwest`;
- `3` = `Move Southeast`;
- `7` = `Move Northwest`;
- `9` = `Move Northeast`.

# New key bindings:

- `Shift` + `Cursor Left` -> `7` = `Move Northwest`;
- `Shift` + `Cursor Right` -> `3` = `Move Southeast`;
- `Shift` + `Cursor Up` -> `9` = `Move Northeast`;
- `Shift` + `Cursor Down` -> `1` =  `Move Southwest`.

and

- `Ctrl` + `Cursor Left` -> `1` = `Move Southwest`;
- `Ctrl` + `Cursor Right` -> `9` = `Move Northeast`;
- `Ctrl` + `Cursor Up` -> `7` = `Move Northwest`;
- `Ctrl` + `Cursor Down` -> `3` = `Move Southeast`.

# Illustration:

Illustration

# Screenshots:

default

### CODE REVIEW

**Comment by ZhilkinSerg** [comment]
jenkins rebuild

File: src/sdltiles.cpp [hunk 1]
```diff
@@ -1131,10 +1131,38 @@ int HandleDPad()
...
+ * or something that a call to ncurses getch would return.
  */
 long sdl_keysym_to_curses(SDL_Keysym keysym)
 {
+
+    //Shift + Arrow (diagonal clockwise)
+    if( keysym.mod & KMOD_SHIFT ){
+        switch (keysym.sym) {
+            case SDLK_LEFT:
+                return '7'; //LEFTUP;
```
Comment (codemime) [inline src/sdltiles.cpp]: Numeric constants in `src/input.h` would be better than the magic numbers.

File: src/sdltiles.cpp [hunk 2]
```diff
@@ -1131,10 +1131,38 @@ int HandleDPad()
...
+                return '3'; //RIGHTDOWN;
+            case SDLK_UP:
+                return '9'; //RIGHTUP;
+            case SDLK_DOWN:
+                return '1'; //LEFTDOWN;
+        }
+    }
+    //Ctrl + Arrow (diagonal counter-clockwise)
+    if( keysym.mod & KMOD_CTRL ){
+        switch (keysym.sym) {
```
Comment (codemime) [inline src/sdltiles.cpp]: Formatting.

**Comment by codemime** [comment]
Any luck with `ncurses`?

**Comment by AMurkin** [comment]
Illustration:
image

File: src/sdltiles.cpp (hunk 2 above)
Comment (ZhilkinSerg) [reply src/sdltiles.cpp]: Thanks. Astyled my changes.

File: src/sdltiles.cpp (hunk 1 above)
Comment (ZhilkinSerg) [reply src/sdltiles.cpp]: I agree. My final aim is to be able to load keys from keybindings.

**Comment by ZhilkinSerg** [comment]
> Any luck with `ncurses`?

Didn't look into `ncurses` yet.

@AMurkin, thanks for the picture.

**Comment by BevapDin** [comment]
> My final aim is to be able to load keys from keybindings.

In that case, you should probably change "input.cpp" and add support for modifier keys. `input_event` already has an (unused) member `modifiers`. This should be filled with the status of the shift/control/(maybe)alt keys.

At least for SDL builds, the modifiers can be set in "sdltiles.cpp" (maybe where `last_input` is set `input_event( the_key, CATA_INPUT_KEYBOARD)`). You'll have to change functions like `popup_getkey` to return an `input_event` instead of the plain character to get that information into the key binding setup in "input.cpp". You'll also have to enhance the loading of keybindings (`input_manager::load`) to load the modifiers from JSON.

File: src/sdltiles.cpp (hunk 1 above)
Comment (ZhilkinSerg) [reply src/sdltiles.cpp]: Movement keys are now taken from input manager.

**Comment by ZhilkinSerg** [comment]
> In that case, you should probably change "input.cpp" and add support for modifier keys. input_event already has an (unused) member modifiers. This should be filled with the status of the shift/control/(maybe)alt keys.
At least for SDL builds, the modifiers can be set in "sdltiles.cpp" (maybe where last_input is set input_event( the_key, CATA_INPUT_KEYBOARD)). You'll have to change functions like popup_getkey to return an input_event instead of the plain character to get that information into the key binding setup in "input.cpp". You'll also have to enhance the loading of keybindings (input_manager::load) to load the modifiers from JSON.

Thanks, but it happened to be too hard for me to implement this, so I decided not to save modifiers together with keybindings, but simply read action characters from `input_manager`.

**Comment by ZhilkinSerg** [comment]
Should be good to go for `sdl` builds.

I've started research for `ncurses` support in separate branch.

File: src/input.cpp [hunk 3]
```diff
@@ -419,6 +419,16 @@ const std::vector<input_event> &input_manager::get_input_for_action( const std::
     return attributes.input_events;
 }
 
+char input_manager::get_first_char_for_action( const std::string &action_descriptor,
+        const std::string context )
+{
+    const input_event first_input_event = get_input_for_action( action_descriptor, context )[0];
+    long first_input = first_input_event.get_first_input();
+    char first_input_char = ( char )first_input;
```
Comment (BevapDin) [inline src/input.cpp]: Why are you explicitly casting this using a C-style cast (there is already an implicit conversion for the assignment)?

Why the local variable, why not return it directly?

Why return a `char`, when the return value (`first_input`) is a `long` *and* the result of the function is promoted to a `long` again (see return statement in `sdl_keysym_to_curse`, which returns a `long`)?

File: src/input.cpp [hunk 4]
```diff
@@ -419,6 +419,16 @@ const std::vector<input_event> &input_manager::get_input_for_action( const std::
     return attributes.input_events;
 }
 
+char input_manager::get_first_char_for_action( const std::string &action_descriptor,
+        const std::string context )
+{
+    const input_event first_input_event = get_input_for_action( action_descriptor, context )[0];
```
Comment (BevapDin) [inline src/input.cpp]: How can you be sure that the vector returned by `get_input_for_action` always has at least one entry?

File: src/input.h [hunk 5]
```diff
@@ -181,6 +181,12 @@ class input_manager
         const std::vector<input_event> &get_input_for_action( const std::string &action_descriptor,
                 const std::string context = "default", bool *overwrites_default = NULL );
 
+        /**
+         * Return first char associated with an action ID in a given context.
+         */
+        char get_first_char_for_action( const std::string &action_descriptor,
+                                        const std::string context = "default" );
```
Comment (BevapDin) [inline src/input.h]: Why is there a default value for the second parameter? All calls to this function supply a specific value anyway, the default is *never* used. And why is it a const value, but the first parameter is a reference to const?

File: src/sdltiles.cpp [hunk 6]
```diff
@@ -1131,11 +1131,41 @@ int HandleDPad()
...
+long sdl_keysym_to_curses( SDL_Keysym keysym )
 {
-    switch (keysym.sym) {
+
+    if( get_option<bool>( "DIAG_MOVE_WITH_MODIFIERS" ) ) {
+        //Shift + Cursor Arrow (diagonal clockwise)
+        if( keysym.mod & KMOD_SHIFT ) {
+            switch( keysym.sym ) {
+                case SDLK_LEFT:
+                    return inp_mngr.get_first_char_for_action( "LEFTUP", "keyboard" );
```
Comment (BevapDin) [inline src/sdltiles.cpp]: I can't find any input context "keyboard", did you mean "default"?

File: src/options.cpp [hunk 7]
```diff
@@ -1030,6 +1030,13 @@ void options_manager::init()
 
     mOptionsSort["interface"]++;
 
+    add( "DIAG_MOVE_WITH_MODIFIERS", "interface", translate_marker( "Diagonal movement with cursor keys and modifiers" ),
+        translate_marker( "If true, allows diagonal movement with cursor keys using CTRL and SHIFT modifiers (<color_red>currently only in SDL builds</color>).  Diagonal movement action keys are taken from keybindings, so you need these to be configured." ),
```
Comment (BevapDin) [inline src/options.cpp]: No other build specific option mentions that is only available in certain build types. (Note that `COPT_CURSES_HIDE` already hides it on curses builds.) For the sake of consistency, this should not be mentioned at all, or be mentioned by all options, preferably automated by checking `cOpt::hide`.

File: src/input.cpp (hunk 3 above)
Comment (ZhilkinSerg) [reply src/input.cpp]: That is true. I've simplified new function to remove redundant type cast.

File: src/input.cpp (hunk 4 above)
Comment (ZhilkinSerg) [reply src/input.cpp]: I've added vector size check.

File: src/sdltiles.cpp (hunk 6 above)
Comment (ZhilkinSerg) [reply src/sdltiles.cpp]: Context is unnecessary, so I removed it.

File: src/options.cpp (hunk 7 above)
Comment (ZhilkinSerg) [reply src/options.cpp]: I've removed SDL build mention from option description.

File: src/input.h (hunk 5 above)
Comment (ZhilkinSerg) [reply src/input.h]: That was copied from `get_input_for_action`.

**Comment by ZhilkinSerg** [comment]
> Any luck with `ncurses`?

I have to admit `ncurses` is out of my expertise now.

**Comment by codemime** [comment]
Works as expected (although, it took some time to get used to it). I'd only change two things:
- The menu option shouldn't appear in ncurses builds since it doesn't affect them for now;
- I'd make the option default in SDL builds (at first I thought the feature didn't work).

**Comment by ZhilkinSerg** [comment]
> The menu option shouldn't appear in ncurses builds since it doesn't affect them for now;

There is `COPT_CURSES_HIDE` modifier, so option won't appear in CURSES builds (similar to `USE_TILES` option).

````CPP
    add( "DIAG_MOVE_WITH_MODIFIERS", "interface", translate_marker( "Diagonal movement with cursor keys and modifiers" ),
        translate_marker( "If true, allows diagonal movement with cursor keys using CTRL and SHIFT modifiers.  Diagonal movement action keys are taken from keybindings, so you need these to be configured." ),
        true, COPT_CURSES_HIDE
        );
````

> I'd make the option default in SDL builds (at first I thought the feature didn't work).

I've changed default value of `DIAG_MOVE_WITH_MODIFIERS` to `true`.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/dropping-things-on-the-wrong-tile-constantly/18223/8

**Pull request merged**